machine/xen/v1alpha1.go

```diff
@@ -0,0 +1,536 @@
+						Hostname: iface.Spec.Hostname,
+						Domain:   iface.Spec.Domain,
+					}),
+				)
+				i++
+			}
+		}
+	}
+
+	var fstab []string
+
+	// TODO(andreistan26): Check if installed xen supports 9pfs
+	for i, vol := range machine.Spec.Volumes {
+		switch vol.Spec.Driver {
+		case "9pfs":
+			mounttag := fmt.Sprintf("fs%d", i+1)
+			p9Dev, err := xenlight.NewDeviceP9()
+			if err != nil {
+				return nil, err
+			}
+
+			p9Dev.Tag = mounttag
+			p9Dev.Path = vol.Spec.Source
+			p9Dev.SecurityModel = "none"
+
+			cfg.P9S = append(cfg.P9S, *p9Dev)
+
+			fstab = append(fstab, vfscore.NewFstabEntry(
+				mounttag,
+				vol.Spec.Destination,
+				vol.Spec.Driver,
+				"",
+				"",
+				"",
+			).String())
+		default:
+			return machine, fmt.Errorf("unsupported Xen volume driver: %v", vol.Spec.Driver)
+		}
+	}
+
+	if len(fstab) > 0 {
+		kernelArgs = append(kernelArgs, vfscore.ParamVfsFstab.WithValue(fstab))
+	}
+
+	args := kernelArgs.Strings()
+	if len(args) > 0 {
+		args = append(args, "--")
+	}
+
+	args = append(args, machine.Spec.ApplicationArgs...)
+	cfg.BInfo.Cmdline = strings.Join(args, " ")
+
+	xenCtx, err := xenlight.NewContext()
+	if err != nil {
+		return nil, fmt.Errorf("could not create xen context: %v", err)
+	}
+	defer xenCtx.Close()
+
+	machine.CreationTimestamp = metav1.Now()
+
+	domID, err := xenCtx.DomainCreateNew(cfg)
+	if err != nil {
+		return machine, fmt.Errorf("could not create xen domain: %v", err)
+	}
+
+	machine.Status.PlatformConfig = domID
+
+	machine.Status.State = machinev1alpha1.MachineStateCreated
+
+	return machine, nil
+}
+func (service *machineV1alpha1Service) Start(ctx context.Context, machine *machinev1alpha1.Machine) (*machinev1alpha1.Machine, error) {
+	if machine.Status.PlatformConfig == nil {
+		return machine, fmt.Errorf("machine has no platform config")
+	}
+
+	domId, ok := machine.Status.PlatformConfig.(xenlight.Domid)
+	if !ok {
+		return machine, fmt.Errorf("machine has no platform config")
+	}
+
+	xenCtx, err := xenlight.NewContext()
+	if err != nil {
+		return nil, fmt.Errorf("could not create xen context: %v", err)
+	}
+	defer xenCtx.Close()
+
+	err = xenCtx.DomainUnpause(domId)
+	if err != nil {
+		return machine, fmt.Errorf("could not unpause xen domain: %v", err)
+	}
+
+	machine.Status.State = machinev1alpha1.MachineStateRunning
+	machine.Status.StartedAt = time.Now()
+
+	return machine, nil
+}
+func (service *machineV1alpha1Service) Pause(ctx context.Context, machine *machinev1alpha1.Machine) (*machinev1alpha1.Machine, error) {
+	if machine.Status.PlatformConfig == nil {
+		return machine, fmt.Errorf("machine has no platform config")
+	}
+
+	domId, ok := machine.Status.PlatformConfig.(xenlight.Domid)
+	if !ok {
+		return machine, fmt.Errorf("machine has no platform config")
+	}
+
+	xenCtx, err := xenlight.NewContext()
+	if err != nil {
+		return nil, fmt.Errorf("could not create xen context: %v", err)
+	}
+	defer xenCtx.Close()
+
+	if err := xenCtx.DomainUnpause(domId); err != nil {
```

Comment:
Why does `Pause` do `Unpausing` is it a typo?

Reply:
Nice catch, forgot to test the `pause` command. That should be `DomainPause`.